xbmc/FileItem.cpp

```diff
@@ -2870,289 +2873,225 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
             }
-            else // Title mismatch
-            {
-              offset = 0;
-              ++expr;
-              break;
-            }
-          }
-          else // No match 2, next expression
-          {
-            offset = 0;
-            ++expr;
-            break;
+            if (stack.size() > 1)
+              deleteItems.emplace_back(index); // delete all but first
           }
-          j++;
         }
-        if (j == Size())
-          expr = stackRegExps.end();
-      }
-      else // No match 1
-      {
-        offset = 0;
-        ++expr;
-      }
-      if (stack.size() > 1)
-      {
-        // have a stack, remove the items and add the stacked item
-        // dont actually stack a multipart rar set, just remove all items but the first
-        std::string stackPath;
+
+        // Create stack
+        // Generate combined stack path
+        std::string stackPath{};
         if (Get(stack[0])->IsRAR())
           stackPath = Get(stack[0])->GetPath();
         else
-        {
-          CStackDirectory dir;
-          stackPath = dir.ConstructStackPath(*this, stack);
-        }
-        item1->SetPath(stackPath);
-        // clean up list
-        for (unsigned k = 1; k < stack.size(); k++)
-          Remove(i+1);
-        // item->m_bIsFolder = true;  // don't treat stacked files as folders
-        // the label may be in a different char set from the filename (eg over smb
-        // the label is converted from utf8, but the filename is not)
-        if (!CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_FILELISTS_SHOWEXTENSIONS))
+          stackPath = CStackDirectory::ConstructStackPath(*this, stack);
+
+        // First item in stack becomes the stack
+        std::string stackName = std::get<1>(stackBaseItem); // title
+        if (!CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+                CSettings::SETTING_FILELISTS_SHOWEXTENSIONS))
           URIUtils::RemoveExtension(stackName);
 
-        item1->SetLabel(stackName);
-        item1->m_dwSize = size;
-        break;
+        // Update item
+        CFileItemPtr stackItem{Get(stack[0])};
+        stackItem->SetPath(stackPath);
+        stackItem->SetLabel(stackName);
+        stackItem->m_dwSize = size;
+      }
+    }
+
+    // Delete unneeded items
+    std::sort(deleteItems.begin(), deleteItems.end(),
+              [](const int i, const int j) { return i > j; });
+    for (int i : deleteItems)
+      Remove(i);
+
+    // Update non-stack items
+    for (const auto& item : m_items)
+    {
+      const std::string playPath = item->GetOpticalMediaPath();
+      if (!item->IsStack() && !playPath.empty())
+      {
+        item->SetPath(playPath); // updated path (for DVD/Bluray files)
+        item->m_bIsFolder = false;
+        item->SetLabel2("");
+        item->SetLabelPreformatted(true);
+        m_sortDescription.sortBy = SortByNone; // sorting is now broken
```

Comment:
Not sure whether rewriting m_items' item paths does have bad side effects.

Reply:
This was in original code. In fairness it's completely unrelated to stacks. I'll have a look and see if I can move it before - as SetPath is called in a couple of places for this reason (one for stacks, one for those not stacked).